release-24.1: stmtbundle: refactor FK handling in schema.sql

yuzefovich opened the pull request:
Backport 4/4 commits from #142151.

/cc @cockroachdb/release

---

This PR refactors how we include tables and FKs into the bundles. It also optimizes _which_ tables we include. See each commit for details.

Fixes: #139196.
Fixes: #142006.

Release justification: low-risk supportability improvement.

yuzefovich committed:
sql: add IGNORE_FOREIGN_KEYS option to SHOW CREATE TABLE

This will be used in the following commit for stmt bundle construction.

`SHOW CREATE TABLE` command has been extended with `WITH IGNORE_FOREIGN_KEYS`
option which makes it so that we use existing `create_nofks` column of
`crdb_internal.create_statements` virtual table. This new option cannot
be combined with existing `WITH REDACT` option (because we don't have
the corresponding column, at least yet).

Release note (sql change): New `WITH IGNORE_FOREIGN_KEYS` option is
added to `SHOW CREATE TABLE` stmt which makes it so that the foreign key
constraints are not included in the output schema. This option is also
acceptable in `SHOW CREATE VIEW` but has no influence there. It cannot
be combined with existing `WITH REDACT` option.

yuzefovich committed:
stmtbundle: refactor FK handling in schema.sql

As of 982abc239efa69a808e4a2b4ecbb01126b253d8f, we include all FK
referenced and referencing tables, including transitively, into the stmt
bundles for all queries. This information might be required to
understand the plans in some cases (like when we have a mutation which
results in a CASCADE post-query), so we initially went with "include
everything all the time". This can be suboptimal since we could pull in
information that isn't actually needed for a particular query.
Additionally, we've just seen a case where in the multi-region cluster
pulling in all these related tables was taking minutes.

To partially alleviate this problem, this commit refactors how we
include FKs into the bundles. Namely, unless a redacted bundle is
requested, we'll use newly-added `WITH IGNORE_FOREIGN_KEYS` option of
`SHOW CREATE TABLE`; then we'll add only relevant FKs as ALTER TABLE
statements. For read-only queries we now include referenced by metadata
tables as well as FKs between them (i.e. both origin and referenced
tables of a FK constraint must be referenced by the metadata to
be included); for mutations we still include all referenced and
referencing tables, including transitively, as well as all FKs. Reducing
the set of tables for mutations is done in the following commit.

As a bonus, this refactor allows us to gracefully handle FK cycles (i.e.
the bundles will be recreatable).

Release note: None

yuzefovich committed:
stmtbundle: optimize FK inclusion for mutations

This commit further optimizes which FK reference tables we include into
the stmt bundle for mutation statements. We always include all tables
that are referenced by the metadata. Then for each foreign key we need
to decide on two actions:
1) whether to include the reference table, and
2) whether FKs of the reference table should be "explored".

This commit implements the following logic using the newly-extracted
visitor method:
- if we're visiting a referenced table of a FK. This means that the
origin table has already been visited (i.e. we visited `child` and are
visiting `parent`). Only for DELETEs we don't want to include the
`parent` because it is unaffected, for all other mutations we do need to
perform the FK check against `parent` when modifying `child`. (Note that,
strictly speaking, we only need to include `parent` if the FK columns in
`child` are actually modified by the stmt, but it seems tricky to get
that right, so we don't try.)
- if we're visiting an origin table of a FK. This means that the
referenced table has already been visited (i.e. we visited `parent`
and are visiting `child`).
  - if we have an INSERT into `parent`, then we don't need to include
`child`
  - if we have DELETE, UPDATE, or UPSERT, then we always will need to
 perform the FK check, so we always include `child`.
    - additionally, if we have a relevant CASCADE action, we also want
to explore `child`'s FKs since we might execute a CASCADE postquery
which could trigger more FK checks / cascades.

Note that some of the FK reference tables are already referenced by the
metadata, so we might end up visiting an extra "layer" of FK references.
Avoiding that seems quite tricky, so we don't do that.

Worth noting that we also have an imprecise way of determining whether
a particular mutation type takes place (i.e. whether we have an INSERT
or not), based on the plan flags that are added in this commit and
describe the whole plan.

This commit additionally adds an escape hatch - a cluster setting that
allows us to include all FK references unconditionally.

Release note: None

yuzefovich commented:
The conflicts here were also minor (needed to rewrite `explain_env` in addition to go import deletion and missing plan flag).

blathers backport release-23.2

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142164)
<!-- Reviewable:end -->

yuzefovich committed:
sql: refactor table name retrieval with FK reference tables

This commit refactors how we fetch "all data source names" that we need
for stmt bundles and EXPLAIN (OPT, ENV). It effectively inlines
`AllDataSourceNames` method in two callsites and extracts
`Metadata.getAllReferenceTables` into an exported visitor method.

The new visitor method is additionally extended with the following
abilities:
- it takes two callbacks:
  - `visitPreFn` controls whether the visitor explores the FK reference
tables of the given table (i.e. whether the visitor should "recurse" into
FK reference tables of the given one)
  - `visitFn` allows the caller to do some work on the table (like
including into a list).
- a particular table is now considered "fully processed" once it's
explored. This means that the same table can be visited multiple
times.

These abilities aren't utilized yet, so this commit is a no-op for now
(i.e. there is no behavior change). This commit results in some minor
code duplication but removes some tight coupling.

Release note: None
